mentat/session_stream.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations
+
+import asyncio
+import logging
+from contextvars import ContextVar
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum
+from typing import Any, AsyncGenerator, Dict, List, cast
+from uuid import UUID, uuid4
+
+from .broadcast import Broadcast
+
+logger = logging.getLogger()
+
+_SESSION_STREAM: ContextVar[SessionStream] = ContextVar("mentat:session_stream")
+
+
+def get_session_stream():
+    return _SESSION_STREAM.get()
+
+
+def set_session_stream(session_stream: SessionStream):
+    return _SESSION_STREAM.set(session_stream)
+
+
+class StreamMessageSource(Enum):
+    SERVER = "server"
+    CLIENT = "client"
+
+
+@dataclass
+class StreamMessage:
+    id: UUID
+    channel: str
+    source: StreamMessageSource
+    data: Any
+    extra: Dict[str, Any] | None
+    created_at: datetime
+
+
+class SessionStream:
+    """Replaces `cprint` and `print`
+
+    Stores message history for a Session and holds an in-memory message bus.
+
+    Terminal and extension clients can read these messages and render them accordingly.
+    For the terminal, they would be rendered with `cprint`.
+    """
+
+    def __init__(self):
+        self.messages: List[StreamMessage] = []
+        self.interrupt_lock = asyncio.Lock()
```

Comment:
This feels janky, will come up with a better way to handle interrupts from clients later.